### PR TITLE
Update process-reboot-cause to save the reboot-cause to state-db 

### PR DESCRIPTION
**Pull request by sujinmkang**
**- Why I did it**
To make the reboot cause history information available for telemetry service, during processing the reboot-cause on boot, the information is saved in state-db.
**- How I did it**
Split the `process-reboot-cause service` into two services: `determine-reboot-cause` and `process-reboot-cause`
1. determine-reboot-cause :
 Determine the reboot-cause based on the hardware/software reboot-cause information and save it into a file with time stamp when it's determined.
2. process-reboot-cause : 
 Read the saved reboot-cause files up to 10 files (after time-sorted) from "/host/reboot-cause/history" directory and save them to state-db. 
**- How to verify it**
Tested in DUT using Cli  commands ( introduced and modified with this PR : https://github.com/Azure/sonic-utilities/pull/1154) and check the state-db with redis-cli.

**- Which release branch to backport (provide reason below if selected)**

- [ ] 201811
- [ ] 201911
- [ ] 202006

**- Description for the changelog**

HLD : https://github.com/Azure/SONiC/pull/669

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Review by jleveque — COMMENTED** [review]
If we go with this approach, we might want to consider naming the first service "determine-reboot-cause" and the second service "process-reboot-cause", or similar. It's more meaningful than "post-process-reboot-cause".

**Review by jleveque — CHANGES_REQUESTED** [review]
Now that there are two services, I suggest renaming the directory from `files/image_config/process-reboot-cause` to `files/image_config/reboot-cause`

**Comment by sujinmkang** [comment]
retest mellanox!

**Comment by sujinmkang** [comment]
@jleveque Can you review this again?

File: files/image_config/reboot-cause/determine-reboot-cause [hunk 1]
```diff
@@ -153,6 +222,9 @@ def main():
     #    reboot info. We will use it as the previous cause.
     software_reboot_cause = find_software_reboot_cause()
 
+    # 4. If there is any additional platform info to be stored for reboot cause analysis
+    #additional_reboot_info = find_platform_reboot_info()
```
[the remark this answers is not in the document]
Comment (sujinmkang) [reply files/image_config/reboot-cause/determine-reboot-cause:227]: Added TODO for later update with additional support if needed.

File: files/image_config/reboot-cause/process-reboot-cause [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+except ImportError as err:
+    raise ImportError("%s - required module not found" % str(err))
+
+VERSION = "1.0"
+
+SYSLOG_IDENTIFIER = "process-reboot-cause"
+
+REBOOT_CAUSE_DIR = "/host/reboot-cause/"
+PREVIOUS_REBOOT_CAUSE_DIR = "/host/reboot-cause/previous-reboot-cause/"
+PREVIOUS_REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "previous-reboot-cause.txt"
```
Comment (jleveque) [inline files/image_config/reboot-cause/process-reboot-cause:27]: As per my comments on the HLD, this file now seems redundant. I suggest we eliminate it altogether.

File: files/image_config/reboot-cause/determine-reboot-cause [hunk 3]
```diff
@@ -1,26 +1,31 @@
...
 
 VERSION = "1.0"
 
-SYSLOG_IDENTIFIER = "process-reboot-cause"
+SYSLOG_IDENTIFIER = "determine-reboot-cause"
 
 REBOOT_CAUSE_DIR = "/host/reboot-cause/"
+PREVIOUS_REBOOT_CAUSE_DIR = "/host/reboot-cause/previous-reboot-cause/"
 REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "reboot-cause.txt"
 PREVIOUS_REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "previous-reboot-cause.txt"
```
Comment (jleveque) [inline files/image_config/reboot-cause/determine-reboot-cause:30]: As per my comments on the HLD, this file now seems redundant. I suggest we eliminate it altogether.

File: files/image_config/reboot-cause/determine-reboot-cause [hunk 4]
```diff
@@ -121,6 +126,67 @@ def find_hardware_reboot_cause():
...
+    else:
+        log_info("No additional reboot info from platform api")
+
+    return additional_reboot_info
+
+
+def create_prev_cause_history_file(previous_reboot_cause, comment):
+    # Create previous_reboot_cause_#time#.txt under previous_reboot_cause directory
+    FILE_GEN_TIME = str(datetime.datetime.now().strftime('%Y_%m_%d_%H_%M_%S'))
+    HIST_FILE = 'previous-reboot-cause-' + FILE_GEN_TIME + '.txt'
```
Comment (jleveque) [inline files/image_config/reboot-cause/determine-reboot-cause:157]: As per my comments on the HLD, I think we can shorten this prefix to "reboot-cause-", since there will be multiple files with this prefix and they will each have a timestamp. "previous" no longer has any significance, and instead could be more confusing.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: files/image_config/reboot-cause/determine-reboot-cause (hunk 3 above)
Comment (sujinmkang) [reply files/image_config/reboot-cause/determine-reboot-cause:30]: this is directory name not file name.

File: files/image_config/reboot-cause/determine-reboot-cause [hunk 5]
```diff
@@ -1,28 +1,33 @@
...
 VERSION = "1.0"
 
-SYSLOG_IDENTIFIER = "process-reboot-cause"
+SYSLOG_IDENTIFIER = "determine-reboot-cause"
 
 REBOOT_CAUSE_DIR = "/host/reboot-cause/"
+PREVIOUS_REBOOT_CAUSE_DIR = "/host/reboot-cause/previous-reboot-cause/"
 REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "reboot-cause.txt"
-PREVIOUS_REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "previous-reboot-cause.txt"
+PREVIOUS_REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "last-reboot-cause.json"
```
Comment (jleveque) [inline files/image_config/reboot-cause/determine-reboot-cause:30]: Prefer sticking with "previous-" rather than "last-", as it's a bit more specific, whereas "last" can have a few meanings.

File: files/image_config/reboot-cause/process-reboot-cause [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    sonic_logger.log_info("Starting up...")
+
+    if not os.geteuid() == 0:
+        sonic_logger.log_error("User {} does not have permission to execute".format(pwd.getpwuid(os.getuid()).pw_name))
+        sys.exit("This utility must be run as root")
+
+    # Set a default previous reboot cause
+    last_reboot_cause = REBOOT_CAUSE_UNKNOWN
+
+    # Read the last previous reboot cause
```
Comment (jleveque) [inline files/image_config/reboot-cause/process-reboot-cause:81]: Suggest rewording: `# Read the most recent reboot cause file and log data to syslog`

File: files/image_config/reboot-cause/process-reboot-cause [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+except ImportError as err:
+    raise ImportError("%s - required module not found" % str(err))
+
+VERSION = "1.0"
+
+SYSLOG_IDENTIFIER = "process-reboot-cause"
+
+REBOOT_CAUSE_DIR = "/host/reboot-cause/"
+PREVIOUS_REBOOT_CAUSE_DIR = "/host/reboot-cause/previous-reboot-cause/"
+PREVIOUS_REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "last-reboot-cause.json"
```
Comment (jleveque) [inline files/image_config/reboot-cause/process-reboot-cause:27]: Prefer sticking with "previous-" rather than "last-", as it's a bit more specific, whereas "last" can have a few meanings.

File: files/image_config/reboot-cause/determine-reboot-cause [hunk 8]
```diff
@@ -1,28 +1,33 @@
...
 except ImportError as err:
     raise ImportError("%s - required module not found" % str(err))
 
 VERSION = "1.0"
 
-SYSLOG_IDENTIFIER = "process-reboot-cause"
+SYSLOG_IDENTIFIER = "determine-reboot-cause"
 
 REBOOT_CAUSE_DIR = "/host/reboot-cause/"
+PREVIOUS_REBOOT_CAUSE_DIR = "/host/reboot-cause/previous-reboot-cause/"
```
Comment (jleveque) [inline files/image_config/reboot-cause/determine-reboot-cause:28]: Suggest changing name of directory:

```
REBOOT_CAUSE_HSITORY_DIR = "/host/reboot-cause/history/"
```

File: files/image_config/reboot-cause/process-reboot-cause [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    from sonic_py_common import device_info, logger
+except ImportError as err:
+    raise ImportError("%s - required module not found" % str(err))
+
+VERSION = "1.0"
+
+SYSLOG_IDENTIFIER = "process-reboot-cause"
+
+REBOOT_CAUSE_DIR = "/host/reboot-cause/"
+PREVIOUS_REBOOT_CAUSE_DIR = "/host/reboot-cause/previous-reboot-cause/"
```
Comment (jleveque) [inline files/image_config/reboot-cause/process-reboot-cause:26]: Suggest changing name of directory:

```
REBOOT_CAUSE_HSITORY_DIR = "/host/reboot-cause/history/"
```

**Review by jleveque — CHANGES_REQUESTED** [review]

File: files/image_config/reboot-cause/determine-reboot-cause [hunk 10]
```diff
@@ -121,6 +126,67 @@ def find_hardware_reboot_cause():
...
+    else:
+        log_info("No additional reboot info from platform api")
+
+    return additional_reboot_info
+
+
+def create_prev_cause_history_file(previous_reboot_cause, comment):
+    # Create reboot_cause_#time#.json under previous_reboot_cause directory
+    FILE_GEN_TIME = str(datetime.datetime.now().strftime('%Y_%m_%d_%H_%M_%S'))
+    REBOOT_CAUSE_TIME_FILE = 'reboot-cause-' + FILE_GEN_TIME + '.json'
```
Comment (jleveque) [inline files/image_config/reboot-cause/determine-reboot-cause:157]: Suggest renaming: `REBOOT_CAUSE_TIME_FILE` --> `REBOOT_CAUSE_HISTORY_FILE`

File: files/image_config/reboot-cause/determine-reboot-cause [hunk 11]
```diff
@@ -121,6 +126,67 @@ def find_hardware_reboot_cause():
...
+    REBOOT_CAUSE_HISTORY_FILE = REBOOT_CAUSE_HISTORY_DIR + REBOOT_CAUSE_TIME_FILE
```
Comment (jleveque) [inline files/image_config/reboot-cause/determine-reboot-cause:158]: Suggest renaming: `REBOOT_CAUSE_HISTORY_FILE` --> `REBOOT_CAUSE_HISTORY_FILE_PATH`

Also, suggest forming the path using `os.path.join()`

File: files/image_config/reboot-cause/process-reboot-cause [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    state_db.set(state_db.STATE_DB, key, 'cause', cause)
+    state_db.set(state_db.STATE_DB, key, 'time', time)
+    state_db.set(state_db.STATE_DB, key, 'user', user)
+    state_db.set(state_db.STATE_DB, key, 'comment', comment)
+
+
+def read_reboot_cause_files_and_save_state_db():
+    # Sort the previous reboot cause files by creation time
+    FULL_REBOOT_FILE_LIST = [os.path.join(REBOOT_CAUSE_HISTORY_DIR,i) for i in os.listdir(REBOOT_CAUSE_HISTORY_DIR)]
+    TIME_SORTED_FULL_REBOOT_FILE_LIST = sorted(FULL_REBOOT_FILE_LIST, key=os.path.getmtime, reverse=True)
```
Comment (jleveque) [inline files/image_config/reboot-cause/process-reboot-cause:55]: Suggest removing "FULL" from constant names. Seems unnecessary, as there are no references to a other/partial lists.

File: files/image_config/reboot-cause/process-reboot-cause [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    raise ImportError("%s - required module not found" % str(err))
+
+VERSION = "1.0"
+
+SYSLOG_IDENTIFIER = "process-reboot-cause"
+
+REBOOT_CAUSE_DIR = "/host/reboot-cause/"
+REBOOT_CAUSE_HISTORY_DIR = "/host/reboot-cause/history/"
+PREVIOUS_REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "previous-reboot-cause.json"
+USER_ISSUE_REBOOT_CAUSE_REGEX ="User issued \'{}\' command [User: {}, Time: {}]"
```
Comment (jleveque) [inline files/image_config/reboot-cause/process-reboot-cause:28]: s/USER_ISSUE_REBOOT_CAUSE_REGEX/USER_ISSUED_REBOOT_CAUSE_REGEX/

**Review by jleveque — CHANGES_REQUESTED** [review]

File: files/build_templates/sonic_debian_extension.j2 [hunk 14]
```diff
@@ -407,9 +407,14 @@ sudo cp $BUILD_TEMPLATES/process-reboot-cause.timer $FILESYSTEM_ROOT_USR_LIB_SYS
...
-sudo cp $IMAGE_CONFIGS/process-reboot-cause/process-reboot-cause.service $FILESYSTEM_ROOT_USR_LIB_SYSTEMD_SYSTEM
+sudo cp $IMAGE_CONFIGS/reboot-cause/process-reboot-cause.service $FILESYSTEM_ROOT_USR_LIB_SYSTEMD_SYSTEM
 echo "process-reboot-cause.service" | sudo tee -a $GENERATED_SERVICE_FILE
-sudo cp $IMAGE_CONFIGS/process-reboot-cause/process-reboot-cause $FILESYSTEM_ROOT/usr/bin/
+sudo cp $IMAGE_CONFIGS/reboot-cause/process-reboot-cause $FILESYSTEM_ROOT/usr/bin/
+
+# Copy determine-reboot-cause service files
+sudo cp $IMAGE_CONFIGS/reboot-cause/determine-reboot-cause.service $FILESYSTEM_ROOT_USR_LIB_SYSTEMD_SYSTEM
+echo "determine-reboot-cause.service" | sudo tee -a $GENERATED_SERVICE_FILE
+sudo cp $IMAGE_CONFIGS/reboot-cause/determine-reboot-cause $FILESYSTEM_ROOT/usr/bin/
```
Comment (jleveque) [inline files/build_templates/sonic_debian_extension.j2:417]: Nit-picking: Can you move this section _above_ the "Copy process-reboot-cause service files" section?

File: files/image_config/reboot-cause/determine-reboot-cause [hunk 15]
```diff
@@ -1,28 +1,33 @@
...
 VERSION = "1.0"
 
-SYSLOG_IDENTIFIER = "process-reboot-cause"
+SYSLOG_IDENTIFIER = "determine-reboot-cause"
 
 REBOOT_CAUSE_DIR = "/host/reboot-cause/"
+REBOOT_CAUSE_HISTORY_DIR = "/host/reboot-cause/history/"
 REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "reboot-cause.txt"
-PREVIOUS_REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "previous-reboot-cause.txt"
+PREVIOUS_REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "previous-reboot-cause.json"
```
Comment (jleveque) [inline files/image_config/reboot-cause/determine-reboot-cause:30]: Suggest using `os.path.join()`

File: files/image_config/reboot-cause/determine-reboot-cause [hunk 16]
```diff
@@ -1,28 +1,33 @@
...
     raise ImportError("%s - required module not found" % str(err))
 
 VERSION = "1.0"
 
-SYSLOG_IDENTIFIER = "process-reboot-cause"
+SYSLOG_IDENTIFIER = "determine-reboot-cause"
 
 REBOOT_CAUSE_DIR = "/host/reboot-cause/"
+REBOOT_CAUSE_HISTORY_DIR = "/host/reboot-cause/history/"
 REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "reboot-cause.txt"
```
Comment (jleveque) [inline files/image_config/reboot-cause/determine-reboot-cause:29]: Suggest using `os.path.join()`

File: files/image_config/reboot-cause/process-reboot-cause [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+except ImportError as err:
+    raise ImportError("%s - required module not found" % str(err))
+
+VERSION = "1.0"
+
+SYSLOG_IDENTIFIER = "process-reboot-cause"
+
+REBOOT_CAUSE_DIR = "/host/reboot-cause/"
+REBOOT_CAUSE_HISTORY_DIR = "/host/reboot-cause/history/"
+PREVIOUS_REBOOT_CAUSE_FILE = REBOOT_CAUSE_DIR + "previous-reboot-cause.json"
```
Comment (jleveque) [inline files/image_config/reboot-cause/process-reboot-cause:27]: Suggest using `os.path.join()`

**Review by jleveque — CHANGES_REQUESTED** [review]

File: files/image_config/reboot-cause/process-reboot-cause [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+    sonic_logger.set_min_log_priority_info()
+
+    sonic_logger.log_info("Starting up...")
+
+    if not os.geteuid() == 0:
+        sonic_logger.log_error("User {} does not have permission to execute".format(pwd.getpwuid(os.getuid()).pw_name))
+        sys.exit("This utility must be run as root")
+
+    # Set a default previous reboot cause
+    last_reboot_cause = REBOOT_CAUSE_UNKNOWN
```
Comment (jleveque) [inline files/image_config/reboot-cause/process-reboot-cause:79]: Suggest renaming to `previous_reboot_cause` for consistency.

File: files/image_config/reboot-cause/determine-reboot-cause [hunk 19]
```diff
@@ -121,6 +126,66 @@ def find_hardware_reboot_cause():
...
+        reboot_cause_dict['time'] = "N/A"
+
+    reboot_cause_dict['comment'] = comment
+
+    # Create REBOOT_CAUSE_HISTORY_DIR if it doesn't exist
+    if not os.path.exists(REBOOT_CAUSE_HISTORY_DIR):
+        os.makedirs(REBOOT_CAUSE_HISTORY_DIR)
+
+    # Write the previous reboot cause to REBOOT_CAUSE_HISTORY_FILE as a JSON format
+    with open(REBOOT_CAUSE_HISTORY_FILE, "w") as reboot_cause_time_file:
```
Comment (jleveque) [inline files/image_config/reboot-cause/determine-reboot-cause:182]: Suggest renaming var to `reboot_cause_history_file` for consistency

**Review by jleveque — CHANGES_REQUESTED** [review]

File: files/image_config/reboot-cause/determine-reboot-cause [hunk 20]
```diff
@@ -121,6 +126,66 @@ def find_hardware_reboot_cause():
...
+    FILE_GEN_TIME = str(datetime.datetime.now().strftime('%Y_%m_%d_%H_%M_%S'))
+    REBOOT_CAUSE_HISTORY_FILE = os.path.join(REBOOT_CAUSE_HISTORY_DIR, "reboot-cause-{}.json".format(FILE_GEN_TIME))
+
+    # resultant dictionary 
+    reboot_cause_dict = {} 
+    reboot_cause_dict['gen_time'] = FILE_GEN_TIME
+    if re.search(r'User issued', previous_reboot_cause):
+        # Match with "User issued 'reboot' command [User: arlakshm, Time: Thu Oct  1 20:11:50 UTC 2020]"
+        pattern = 'User issued \'(.*)\' command \[User: (.*), Time: (.*)\]'
+        match = re.search(pattern, previous_reboot_cause)
```
Comment (jleveque) [inline files/image_config/reboot-cause/determine-reboot-cause:165]: It seems like this regex is unnecessary. `find_software_reboot_cause()` currently produces a nicely formatted string for display. But since you're changing things to store a JSON file instead, maybe you should modify `find_software_reboot_cause()` to return a dictionary, then you won't need to use regex here. There's no point in creating a user-readable string if it's not going to be displayed anywhere.

**Review by jleveque — CHANGES_REQUESTED** [review]

File: files/image_config/reboot-cause/determine-reboot-cause (hunk 20 above)
Comment (sujinmkang) [reply files/image_config/reboot-cause/determine-reboot-cause:165]: The Software reboot-cause is from "reboot-cause.txt" which was written by reboot script. `find_software_reboot_cause()` doesn't do any processing to create a user-readable string.

Comment (jleveque) [reply files/image_config/reboot-cause/determine-reboot-cause:165]: True, `find_software_reboot_cause()` just appends "First boot of SONiC ..." to the string if applicable. I guess we can keep this for now, but we might want to change the reboot script to write a "reboot-cause.json" file instead, then we can remove all this regex.

File: files/image_config/reboot-cause/process-reboot-cause (hunk 12 above)
Comment (jleveque) [reply files/image_config/reboot-cause/process-reboot-cause]: Also remove "FULL from "TIME_SORTED_FULL_REBOOT_FILE_LIST"

File: files/image_config/reboot-cause/determine-reboot-cause (hunk 20 above)
Comment (sujinmkang) [reply files/image_config/reboot-cause/determine-reboot-cause:165]: Yes we can work on reboot script part later since we need to work on the reboot script to support some other features but not in this PR.

**Review by jleveque — CHANGES_REQUESTED** [review]
Today I merged new changes to create two new packages, `sonic-host-services` and `sonic-host-services-data`

Can you please move the `determine-reboot-cause` and `process-reboot-cause` scripts into the former and move the service files into the latter?

See here: https://github.com/Azure/sonic-buildimage/pull/5694

**Comment by sujinmkang** [comment]
@jleveque I can open a new PR for the migration to sonic-host-service package after I submitting this PR.

**Comment by sujinmkang** [comment]
retest broadcom please

**Comment by sujinmkang** [comment]
retest broadcom please

**Comment by sujinmkang** [comment]
retest broadcom please

File: src/sonic-host-services-data/debian/sonic-host-services-data.process-reboot-cause.timer [hunk 21]
```diff
@@ -3,7 +3,7 @@ Description=Delays process-reboot-cause until network is stably connected
 
 [Timer]
 OnBootSec=1min 30 sec
-Unit=process-reboot-cause.service
+Unit=sonic-host-services-data.process-reboot-cause.service
```
Comment (jleveque) [inline src/sonic-host-services-data/debian/sonic-host-services-data.process-reboot-cause.timer:6]: I don't believe you need to change this, because the service file should get installed as `process-reboot-cause.service`. Did you test this?

File: src/sonic-host-services/scripts/process-reboot-cause [hunk 22]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python
```
Comment (jleveque) [inline src/sonic-host-services/scripts/process-reboot-cause:1]: Please change shebang to `#!/usr/bin/env python3` and ensure script is Python 3-compliant.

**Review by jleveque — CHANGES_REQUESTED** [review]

**Comment by sujinmkang** [comment]
will open a new PR after rebase.
